### PR TITLE
#1202: Add support for calling remote bricks

### PR DESCRIPTION
**Pull request by umair-pixiebrix**

### CODE REVIEW

File: src/background/executor.ts [hunk 1]
```diff
@@ -480,4 +481,15 @@ export async function executeInOpener(
   });
 }
 
+export const executeOnServer = liftBackground(
+  "EXECUTE_ON_SERVER",
+  async (blockId: string, blockArgs: RenderedArgs) => {
```
Comment (twschiller) [inline src/background/executor.ts:486]: Let's use the RegistryId type from `core.ts` for the blockId

File: src/blocks/combinators.ts [hunk 2]
```diff
@@ -227,6 +228,10 @@ async function runStage(
         return await executeInAll(stage.id, blockArgs, baseOptions);
       }
 
+      case "remote": {
+        return (await executeOnServer(stage.id, blockArgs)).data;
```
Comment (twschiller) [inline src/blocks/combinators.ts:232]: This `data` is the data field of the AxiosResponse? Correct? Shouldn't it be data.data?

File: src/background/executor.ts [hunk 3]
```diff
@@ -480,4 +481,15 @@ export async function executeInOpener(
   });
 }
 
+export const executeOnServer = liftBackground(
```
Comment (twschiller) [inline src/background/executor.ts:484]: We shouldn't have AxiosResponse objects crossing component boundaries. Let's return the actual data returned from the response

Comment (twschiller) [reply src/background/executor.ts:484]: How is the `data.error` property from running on the server handled? See "Response Body (status 200)": https://github.com/pixiebrix/pixiebrix-app/issues/307

**Review by twschiller — CHANGES_REQUESTED** [review]
See comment about not passing AxiosResponses across component boundaries, and the question on the error property

File: src/blocks/combinators.ts (hunk 2 above)
Comment (umair-pixiebrix) [reply src/blocks/combinators.ts:232]: Yeah. I changed in backend as well. Let me revert.

File: src/blocks/combinators.ts [hunk 4]
```diff
@@ -227,6 +229,21 @@ async function runStage(
         return await executeInAll(stage.id, blockArgs, baseOptions);
       }
 
+      case "remote": {
+        const { data, error } = (
+          await executeOnServer(stage.id, blockArgs)
+        ).data;
+        if (error) {
+          // TODO: Determine type of error
```
Comment (BALEHOK) [inline src/blocks/combinators.ts:237]: I'd suggest implementing the functionality or creating an Issue. A `TODO` often just lives in code abandoned and nobody knows how/when to deal with it.

Comment (umair-pixiebrix) [reply src/blocks/combinators.ts]: Good idea @BALEHOK. I have added the link of the issue in the comment. https://github.com/pixiebrix/pixiebrix-extension/issues/1227

File: src/blocks/errors.ts [hunk 5]
```diff
@@ -111,3 +111,13 @@ export class OutputValidationError extends BusinessError {
     this.errors = errors;
   }
 }
+
+export class RemoteExecutionError extends BusinessError {
+  readonly error: OutputUnit[];
+
+  constructor(message: string, error: OutputUnit[]) {
+    super(message);
+    this.name = "PipelineConfigurationError";
```
Comment (twschiller) [inline src/blocks/errors.ts:120]: This name needs to match the class name RemoteExecutionError

File: src/background/executor.ts [hunk 6]
```diff
@@ -480,4 +481,15 @@ export async function executeInOpener(
...
+  "EXECUTE_ON_SERVER",
+  async (blockId: RegistryId, blockArgs: RenderedArgs) => {
+    console.debug(`Running ${blockId} on the server`);
+    return (await getLinkedApiClient()).post("/api/run/", {
```
Comment (twschiller) [inline src/background/executor.ts:488]: Annotate post with a type. E.g., regenerate the swagger.ts so you can refer to the type, or just annotate it yourself here

```
post<{data?: JsonObject, error?: JsonObject}>
```

**Review by twschiller — CHANGES_REQUESTED** [review]
Fix RemoteExecutionError class definition
Add return type to post call

File: src/blocks/errors.ts [hunk 7]
```diff
@@ -111,3 +111,13 @@ export class OutputValidationError extends BusinessError {
     this.errors = errors;
   }
 }
+
+export class RemoteExecutionError extends BusinessError {
+  readonly errors: OutputUnit[];
```
Comment (twschiller) [inline src/blocks/errors.ts:116]: @umair-pixiebrix Let's make this `error: unknown` until we land on the error format. There may be other errors aside from input validation errors

File: src/blocks/combinators.ts [hunk 8]
```diff
@@ -227,6 +229,20 @@ async function runStage(
         return await executeInAll(stage.id, blockArgs, baseOptions);
       }
 
+      case "remote": {
+        const { data, errors } = (
```
Comment (twschiller) [inline src/blocks/combinators.ts:233]: Use singular `error` of type unknown

**Review by twschiller — CHANGES_REQUESTED** [review]
For the error, let's have it be `error: unknown`. There will be other other types of errors aside from input validation errors (which have the outputunit type)

File: src/blocks/errors.ts (hunk 7 above)
Comment (umair-pixiebrix) [reply src/blocks/errors.ts:116]: I have used `JsonObject` @twschiller

**Comment by umair-pixiebrix** [comment]
> For the error, let's have it be `error: unknown`. There will be other other types of errors aside from input validation errors (which have the outputunit type)

Does `JsonObject` work?

**Review by twschiller — APPROVED** [review]

**Comment by twschiller** [comment]
@umair-pixiebrix Could you resolve the merge conflict and then merge?

**Pull request merged**